### PR TITLE
[experimental] LayeredMap which views a sequence of MapLayers

### PR DESCRIPTION
**Pull request by msmouse**
## Description

A multi-versional mapping data structure much like the in memory sparse merkle tree but doesn't do the hashing on nodes.
This is useful for implementing the per-block speculative LRU cache / Active State coming up.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Other (specify) - not used yet

## How Has This Been Tested?

proptest

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by msmouse** [comment]
* **#12722** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/12722?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @msmouse and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: experimental/storage/layered-map/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+/// Similar to `Key`, we require `Value: Clone`, another reason being it's tricky to figure out the
+/// lifetime if `get()` returns a reference to the value -- we simply clone the value.
+pub trait Value: Clone + Send + Sync + 'static {}
+
+#[derive(Debug)]
+struct LayerInner<K: Key, V: Value> {
+    root: NodeRef<K, V>,
+    children: Mutex<Vec<Arc<LayerInner<K, V>>>>,
```
Comment (areshand) [inline experimental/storage/layered-map/src/lib.rs:38]: what is children used for?

File: experimental/storage/layered-map/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    fn is_family(&self, other: &Self) -> bool {
+        self.inner.family == other.inner.family
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct LayeredMap<K: Key, V: Value> {
+    bottom_layer: MapLayer<K, V>,
```
Comment (areshand) [inline experimental/storage/layered-map/src/lib.rs:154]: why do we need the bottom_layer? could we assume the LRU cache contains all the base key, value pairs?

File: experimental/storage/layered-map/src/ref.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::{Arc, Weak};
+
+#[derive(Debug)]
+pub enum Ref<T> {
```
[the remark this answers is not in the document]
Comment (msmouse) [reply experimental/storage/layered-map/src/ref.rs:7]: maybe, when it's out of '/experimental'

File: experimental/storage/layered-map/src/lib.rs (hunk 1 above)
Comment (msmouse) [reply experimental/storage/layered-map/src/lib.rs:38]: same with SMT, there can be branches in the block tree.

File: experimental/storage/layered-map/src/lib.rs (hunk 2 above)
Comment (msmouse) [reply experimental/storage/layered-map/src/lib.rs:154]: Think "FrozenSparseMerkleTree" -- you need to make sure nodes created between bottom layer and top layer don't go away while you read, by holding a strong ref to the bottom layer root.

**Review by areshand — APPROVED** [review]

**Review by zekun000 — APPROVED** [review]

**Pull request merged**